third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -1445,18 +1490,28 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
                 self.require_impl_language(&loc);
                 let ref_kind = ReferenceKind::from_is_mut(*mutable);
                 let target_ty = self.fresh_type_var();
-                let result_ty = self.check_type(
-                    &loc,
-                    &Type::Reference(ref_kind, Box::new(target_ty.clone())),
-                    expected_type,
-                    context,
-                );
+                let ty = Type::Reference(ref_kind, Box::new(target_ty.clone()));
+                let result_ty = self.check_type(&loc, &ty, expected_type, context);
                 let target_exp = self.translate_exp(exp, &target_ty);
                 if self.subs.specialize(&target_ty).is_reference() {
                     self.error(&loc, "cannot borrow from a reference")
                 }
-                let id = self.new_node_id_with_type_loc(&result_ty, &loc);
-                ExpData::Call(id, Operation::Borrow(ref_kind), vec![target_exp.into_exp()])
+                // When the target of this borrow is an immutable reference, we still need to set
+                // the type of id to mutable ref, otherwise, the type info the bytecode will be incorrect.
+                // for instance `let x: &T = &mut y;`, borrow of y should still return mutable ref
+                let id = if expected_type.is_immutable_reference() && *mutable {
+                    self.new_node_id_with_type_loc(&ty, &loc)
+                } else {
+                    self.new_node_id_with_type_loc(&result_ty, &loc)
+                };
+                let mut target_exp =
+                    ExpData::Call(id, Operation::Borrow(ref_kind), vec![target_exp.into_exp()])
+                        .into();
+                // Insert freeze for &mut when the expected type is &
+                if self.insert_freeze {
```

Comment:
It seems odd that we could have a situation where `expected_type` is immutable, we have a mutable `Borrow`, but we don't insert a freeze because `self.insert_freeze` is false.  Is such a situation possible?  If so, what happens?  If not, why do we have 2 different ways of testing?

Reply:
For this code:
```
    fun test_6() {
        let x: u64 = 3;
        *(&mut x: &u64) = 5;
    }
```,  if we don't have this condition, we will generate expression:
```
Freeze(Borrow(Mutable)(x)) = 5;
```
However, regardless of inserting freeze or not, the compiler should always return an error. That's why I asked a question above, maybe we don't need to this flag at the beginning. What do you think? @brmataptos

Reply:
As discussed off-line, a similar test case is currently working in `tests/bytecode-generator/mutate_immutable.{move.exp}` in the current PR, but is not working in `tests/checking/typing/mutate_immutable.{move,exp}`.  Despite the identical names, the latter is the same as V1, while the other one is not.  The passes being run differ between those two tests, and it seems likely to be the reason the error is not caught for the `/checking/` test case.

Reply:
Added a new test function in `mutate_immutable_bytecode.move`